testcases/crypto/ssl3_func.c

```diff
@@ -637,6 +637,209 @@ CK_RV do_SSL3_MultipleKeysDerive(CK_SESSION_HANDLE session)
     return rc;
 }
 
+CK_RV do_TLS_PreMasterKeyGen(CK_SESSION_HANDLE session)
+{
+    CK_MECHANISM mech;
+    CK_VERSION version;
+    CK_OBJECT_HANDLE h_key;
+    CK_RV rc = CKR_OK;
+    CK_SLOT_ID slot_id = SLOT_ID;
+
+    testcase_begin("starting do_TLS_PreMasterKeyGen...\n");
+
+    version.major = 3;
+    version.minor = 3;
+
+    mech.mechanism = CKM_TLS_PRE_MASTER_KEY_GEN;
+    mech.pParameter = &version;
+    mech.ulParameterLen = sizeof(CK_VERSION);
+
+    /** skip test if the slot doesn't support this mechanism **/
+    if (!mech_supported(slot_id, mech.mechanism)) {
+        testsuite_skip(1, "Slot %u doesn't support %s (0x%x)",
+                       (unsigned int) slot_id,
+                       mech_to_str(mech.mechanism),
+                       (unsigned int) mech.mechanism);
+        goto done;
+    }
+
+    testcase_new_assertion();
+    rc = funcs->C_GenerateKey(session, &mech, NULL, 0, &h_key);
+    if (rc != CKR_OK) {
+        if (is_rejected_by_policy(rc, session)) {
+            testcase_skip("key generation is not allowed by policy");
+            return CKR_OK;
+        }
+        testcase_fail("C_GenerateKey() rc = %s", p11_get_ckr(rc));
+    } else {
+        testcase_pass("Successfully generated a generic secret key.");
+    }
+
+    if (funcs->C_DestroyObject(session, h_key) != CKR_OK)
+        testcase_error("C_DestroyObject() failed");
+
+done:
+    return rc;
+}
+
+CK_RV do_TLS_MultipleKeysDerive(CK_SESSION_HANDLE session)
+{
+    CK_MECHANISM mech;
+    CK_OBJECT_HANDLE h_pm_secret;
```

Comment:
also here: initialize with "invalid handle" ?

Reply:
this function is an exact copy from above, jus the mechanism is changed. I tend to leave it as is, otherwise I would need to change it in all functions in that file.